src/scripting/game_lua_kernel.hpp

```diff
@@ -57,8 +58,12 @@ class game_lua_kernel : public lua_kernel_base
 	tod_manager & tod_man();
 
 	config level_lua_;
+	int EVENT_TABLE;
+	bool has_preloaded_ = false;
 
 	std::stack<game_events::queued_event const * > queued_events_;
+	
+	utils::guard_value<game_lua_kernel*> resources_ptr_;
```

Comment:
If there are two lua kernels, how do we know which order they'll be destroyed in?

Reply:
There aren't two Lua kernels.

Reply:
Then why do we need to store the old value of `resources::lua_kernel`, instead of knowing that it must be `nullptr`?
And why do it here, instead of doing it in the caller that's creating and destroying the instance of `game_lua_kernel`?

Basically, for every change in a071adc601fa68252713a87c53ff44e0d7bf9512 I'm wondering why it's being done this way.

Reply:
The `resources_ptr_` member is a RAII construct that assigns the Lua kernel pointer to the `resources::lua_kernel` on destruction and nulls out the `resources::lua_kernel` Lua kernel pointer on destruction.

We don't actually _need_ to store the old value of `resources::lua_kernel` since it should always be `nullptr`, but doing it this way allows the `guard_value` class to be useful for other kinds of situations.

Reply:
This doesn't look anything like RAII. It would be a little bit like RAII if `guard_value`'s constructor were to throw when the original value wasn't nullptr, but there's nothing like RAII here at the moment.

My previous comment from April 15th still awaits an answer.

Reply:
That comment doesn't make any sense whatsoever. The `guard_value` class is an RAII class for a scoped temporary change to a variable. It would be downright nonsensical for it to throw based on the prior content of the variable.

I definitely answered your first question. For the second one… putting the `guard_value` in the Lua kernel means that `resources::lua_kernel` is non-null; if and only if the Lua kernel exists. If the `guard_value` was in the kernel's owner, that guarantee would be broken.

Reply:
What is the resource being acquired? If I understand you, the resource is the memory location; however in that case it being non-nullptr means that the resource is already being used - therefore if you're saying this is RAII, it needs to react because the resource is already taken.

If that's not it, then please identify the "resource" in this class' design.

Reply:
> however in that case it being non-nullptr means that the resource is already being used

That doesn't make any sense. It's not even necessarily a pointer. It could be an integer, or a string, or literally anything.

I've already explained its purpose. It's for when you need to _temporarily_ change the value of a variable and want to ensure that the old value gets restored regardless of what exit path is taken. In a sense, it's a special case of the ON_SCOPE_EXIT utility.

I'm not sure there's a "resource" to identify. It could be the memory location, but we're not acquiring exclusive access to it or anything. Or maybe it could be the temporary value stored to that memory location.

Reply:
> I'm not sure there's a "resource" to identify.

You said it's RAII, Resource Acquisition Is Initialisation. The comments of mine that you're saying don't make sense are based on trying to fit it to RAII, because you said it was.

What doesn't make sense is the design of `guard_value` itself.

Above was about the class, below is about the caller.

The way it's being used in the specific case (and in the similar one in the other PR) is to wrap a raw pointer, a non-owning global which is relying on some mechanism to ensure it doesn't become dangling. That's a messy construct which should be redesigned, but putting a temporary-value wrapper around it works in the cases that it already works in, and is still going to break in the cases that it already breaks in. It has a edge-case that it improves (if something else in game_lua_kernel's constructor throws), and an edge-case that it makes worse (if two game_lua_kernels get constructed). On top of that, the class isn't a good fit for the edge-case that it does improve.

Reply:
> That's a messy construct which should be redesigned

Sure, but deprecating the `resources` namespace is extremely out of scope for either PR.

> putting a temporary-value wrapper around it works in the cases that it already works in, and is still going to break in the cases that it already breaks in.

I don't know what you're talking about. By using the `guard_value`, it still works in all the cases that it already worked in, and also works in some cases that it previously broke in. Can you point to _any_ case that would actually break it now?

>  an edge-case that it makes worse (if two game_lua_kernels get constructed)

The only way this could happen is in code of the form:

```cpp
auto lk1 = new game_lua_kernel(...);
auto lk2 = new game_lua_kernel(...);
delete lk1;
delete lk2
```

And I don't think we have a need to support that kind of use-case. That would also break with the old version of this code, so on the whole the `guard_value` improves the situation: it doesn't break anything new, and it makes some things work that previously didn't.

The only thing I'd maybe consider doing differently is making the `guard_value` be a sibling of the `game_lua_kernel` instead of a child. I don't think it makes much difference in practice, however, unless the constructor called something that accessed the kernel via the `resources` namespace.

> On top of that, the class isn't a good fit for the edge-case that it does improve.

I don't agree, but if you have alternative suggestions, I'm listening.

Reply:
Are you now agreeing with me that it's not RAII?

Just leave the code where it was in play_controller.cpp. Leave it unless and until you want to do the proper cleanup in the `resources` namespace.

Reply:
Reverting that change is not an option. The whole reason I did it was to fix a crash, which it did successfully.

Reply:
What are the steps to reproduce the crash, so that I can look at a better way to fix it?

If you want to call this RAII, identify what the R is, otherwise it doesn't make any sense.

Reply:
I don't remember exactly, but I think it was something pretty much impossible to miss, like when you start the scenario, or when an event triggers, or maybe when the scenario ends. I doubt there is a better way to fix it, other than eliminating the use of the `resources` namespace in the offending location.

Reply:
Can you remember any more details about the bug?

I tried reverting that bit of the code, and successfully played my currently-in-progress HttT game from a mid-scenario save through to the second turn of the next scenario. So it doesn't seem to be an impossible-to-miss bug, and I'm interested in whether it indicates a bug elsewhere in this PR.

Reply:
Possibly it only occurs if the scenario adds an event using `wesnoth.game_events.add{}` with at least one of `filter=` or `action=` set to a Lua function? I had added some quick events in the tutorial that do things like that.

Reply:
I wrote a long comment, and think I then clicked on "Resolve" instead of "Comment" afterwards. If you've seen this comment before, that's why it's being reposted; and it's why this got resolved and then unresolved.

The `event_test_filter_lua_serializable` unit test was able to reproduce the bug, and the call-stack of the crash is:
```
#0  in game_lua_kernel::clear_wml_event(int) (with this==nullptr)
#1  in game_events::filter_lua::~filter_lua()
#13 in std::vector<std::shared_ptr<game_events::event_filter>, std::allocator<std::shared_ptr<game_events::event_filter> > >::~vector()
#14 in game_events::event_handler::~event_handler()
#29 in game_events::event_handlers::~event_handlers()
#32 in game_events::manager::~manager()
#35 in game_state::~game_state()
#39 in play_controller::~play_controller()
```

With the old code, play_controller's destructor set the `resources::` pointers to nullptr early in the destruction. The game_events::manager and lua kernel are both owned by the game_state class, and what the guard_value change did was to move the `resources::lua_kernel = nullptr` slightly later in the destruction sequence. The old way that all the `resources::` pointers were cleared early in the destruction sequence seems to be a design choice, that from that point onwards the components would be more isolated.

Changing to having the components call each other during the destruction seems a design change. I don't know whether it's a large or a small change, but I think finding out enough to be sure about it has to be part of this PR. Redesigning the `resources::` pointers might not need to be in the scope of this PR, but having enough knowledge about the mechanism to be sure about it has to be in here. I'm worried that this PR is putting in a new design that doesn't mesh with the old one, without a clear description of what the old design was or how they interact.

Can we be sure that none of the other `resources::` pointers are accessed during the destruction? I'm not sure that we can even be certain that clear_wml_event() won't cause them to be accessed - it removes a reference count, but the Lua object might have a finalizer. For example, src/scripting/lua_terrainfilter.cpp has a finalizer (although it seems to be a no-op), and a Lua based filter that accesses another resource in its finalizer seems plausible, maybe to clear a cache.

Reply:
Is the use of `guard_value` something that can be made into a separate PR, or does the rest of this PR depend on it?  If this is the last thing or nearly the last thing that's remaining to be resolved, then it'd probably to be better to get everything else in and then look at this on its own.

Reply:
Resolved because the `guard_value` was removed.